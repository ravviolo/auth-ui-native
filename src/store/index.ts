import { combineReducers, configureStore, PreloadedState } from '@reduxjs/toolkit';

const rootReducer = combineReducers({});

export const setupStore = (preloadedState?: PreloadedState<RootState>) =>
  configureStore({
    reducer: rootReducer,
    preloadedState,
  });

export const store = setupStore();

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
