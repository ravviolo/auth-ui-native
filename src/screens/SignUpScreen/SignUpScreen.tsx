/* eslint-disable no-console */
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { SignUpTemplate } from 'components/templates';
import { RootStackParamList } from 'navigator';
import { useAppDispatch, useAppSelector } from 'store/hooks';
import { registerUser, selectUser } from 'store/user/userSlice';

type Props = NativeStackScreenProps<RootStackParamList, 'SignUp'>;

export const SignUpScreen = ({ navigation }: Props) => {
  const dispatch = useAppDispatch();
  const user = useAppSelector(selectUser);

  const handleNavigate = () => {
    navigation.navigate('Login');
  };

  const handleSignUpFormSubmit = async (email: string, password: string): Promise<void> => {
    const userCredentials = {
      email,
      password,
    };

    await dispatch(registerUser(userCredentials));
  };

  const handleSignUpViaGoogle = (): void => {
    console.log('Registering user via Google');
  };

  const handleSignUpViaLinkedIn = (): void => {
    console.log('Registering user via LinkedIn');
  };

  const handleSignUpViaFacebook = (): void => {
    console.log('Registering user via Facebook');
  };

  return (
    <SignUpTemplate
      footerBtnText="Login"
      footerText="Already a user?"
      headerText="Sign Up"
      testID="signup-template-test-id"
      onPressFacebook={handleSignUpViaFacebook}
      onPressFooterBtn={handleNavigate}
      onPressGoogle={handleSignUpViaGoogle}
      onPressLinkedIn={handleSignUpViaLinkedIn}
      onSubmitSignUpForm={handleSignUpFormSubmit}
    />
  );
};
