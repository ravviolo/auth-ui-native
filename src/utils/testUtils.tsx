import { PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import testRenderer, { TestRendererOptions, ReactTestRenderer } from 'react-test-renderer';
import { AppStore, RootState, setupStore } from 'store';

interface StoreOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = setupStore(preloadedState) }: StoreOptions = {},
  renderOptions: RenderOptions = {}
) => {
  const Wrapper = ({ children }: React.PropsWithChildren): JSX.Element => {
    return <Provider store={store}>{children}</Provider>;
  };

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export const createWithProviders = (
  ui: React.ReactElement,
  { preloadedState = {}, store = setupStore(preloadedState) }: StoreOptions = {},
  testRendererOptions?: TestRendererOptions
): ReactTestRenderer =>
  testRenderer.create(<Provider store={store}>{ui}</Provider>, testRendererOptions);
