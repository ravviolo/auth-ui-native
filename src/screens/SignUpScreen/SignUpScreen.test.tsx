/* eslint-disable no-console */
import { act, fireEvent, screen, within } from '@testing-library/react-native';
import { RootState } from 'store';
import { createWithProviders, renderWithProviders } from 'utils/testUtils';

import { SignUpScreen } from './SignUpScreen';

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: mockNavigate }),
}));

describe('SignUpScreen', () => {
  const testEmail = 'test@test.com';
  const testPassword = 'password';

  it("initially should have no user's data and hold user status as 'idle'", () => {
    const { store } = renderWithProviders(<SignUpScreen />);

    expect(store.getState().user).toEqual({ userData: null, status: 'idle' });
  });

  it("after sign up should save user's data, change user's status to 'complete'", async () => {
    const { store } = renderWithProviders(<SignUpScreen />);

    const template = screen.getByTestId('signup-template-test-id');
    const emailInputField = within(template).getByTestId('input-field-email-test-id');
    const passwordInputField = within(template).getByTestId('input-field-password-test-id');
    const signUpBtn = within(template).getByTestId('btn-submit-signup-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.changeText(passwordInputField, testPassword);
    await act(() => {
      fireEvent.press(signUpBtn);
    });

    const userStore = store.getState().user;

    expect(userStore.status).toBe('complete');
    expect(userStore.userData?.firstname).toBe('Baby');
    expect(userStore.userData?.lastname).toBe('Yoda');
    expect(userStore.userData?.email).toBe(testEmail);
  });

  it("if user exist should update store with new user's data", async () => {
    const preloadedState: RootState = {
      user: {
        status: 'complete',
        userData: {
          avatarUrl: 'url',
          birthday: 'birthday',
          createdAt: 'createdAt',
          email: 'email',
          firstname: 'firstname',
          lastname: 'lastname',
          username: 'username',
          uuid: 'uuid',
        },
      },
    };

    const { store } = renderWithProviders(<SignUpScreen />, { preloadedState });

    const template = screen.getByTestId('signup-template-test-id');
    const emailInputField = within(template).getByTestId('input-field-email-test-id');
    const passwordInputField = within(template).getByTestId('input-field-password-test-id');
    const signUpBtn = within(template).getByTestId('btn-submit-signup-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.changeText(passwordInputField, 'password');
    await act(() => {
      fireEvent.press(signUpBtn);
    });

    const userStore = store.getState().user;

    expect(userStore.userData?.firstname).toBe('Baby');
    expect(userStore.userData?.lastname).toBe('Yoda');
    expect(userStore.userData?.email).toBe(testEmail);
  });

  it("should log 'Logging user via Google' after clicking on Google icon button", () => {
    renderWithProviders(<SignUpScreen />);
    console.log = jest.fn();

    const googleIconBtn = within(
      screen.getByTestId('auth-socials-signup-template-test-id')
    ).getByTestId(/icon-btn-google-test-id/i);

    fireEvent.press(googleIconBtn);

    expect(console.log).toHaveBeenCalledWith('Registering user via Google');
  });

  it("should log 'Logging user via Facebook' after clicking on Facebook icon button", () => {
    renderWithProviders(<SignUpScreen />);
    console.log = jest.fn();

    const facebookIconBtn = within(
      screen.getByTestId('auth-socials-signup-template-test-id')
    ).getByTestId(/icon-btn-facebook-test-id/i);

    fireEvent.press(facebookIconBtn);

    expect(console.log).toHaveBeenCalledWith('Registering user via Facebook');
  });

  it("should log 'Logging user via LinkedIn' after clicking on LinkedIn icon button", () => {
    renderWithProviders(<SignUpScreen />);
    console.log = jest.fn();

    const linkedInIconBtn = within(
      screen.getByTestId('auth-socials-signup-template-test-id')
    ).getByTestId(/icon-btn-linkedin-test-id/i);

    fireEvent.press(linkedInIconBtn);

    expect(console.log).toHaveBeenCalledWith('Registering user via LinkedIn');
  });

  it('should handle page navigation', () => {
    renderWithProviders(<SignUpScreen />);

    const navigateBtn = within(
      screen.getByTestId('auth-footer-signup-template-test-id')
    ).getByTestId('btn-footer-test-id');

    fireEvent.press(navigateBtn);

    expect(mockNavigate).toHaveBeenCalledWith('Login');
  });

  it('should match snapshot', () => {
    const tree = createWithProviders(<SignUpScreen />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
