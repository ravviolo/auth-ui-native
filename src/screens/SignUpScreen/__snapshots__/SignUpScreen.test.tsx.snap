// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpScreen should match snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "backgroundColor": "#fff",
        "flexGrow": 1,
        "paddingHorizontal": 30,
        "paddingVertical": 50,
      }
    }
    testID="signup-template-test-id"
  >
    <View>
      <View
        testID="auth-header-signup-template-test-id"
      >
        <Text
          style={
            Object {
              "color": "#3D3D3D",
              "fontSize": 18,
              "fontWeight": "bold",
              "textTransform": "uppercase",
            }
          }
          testID="auth-heading-test-id"
        >
          Sign Up
        </Text>
      </View>
      <View
        style={
          Object {
            "height": 220,
            "marginVertical": 60,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
              },
              Object {
                "flex": 1,
                "justifyContent": "space-between",
              },
            ]
          }
          testID="signup-form-test-id"
        >
          <View
            testID="email-input-signup-form-test-id"
          >
            <Text
              style={
                Object {
                  "color": "#3D3D3D",
                  "marginBottom": 3,
                }
              }
              testID="input-label-email-test-id"
            >
              Email
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              secureTextEntry={false}
              style={
                Array [
                  Object {
                    "borderColor": "#9E9E9E",
                    "borderRadius": 8,
                    "borderWidth": 1,
                    "height": 40,
                    "padding": 10,
                  },
                  Object {
                    "borderColor": "#9E9E9E",
                  },
                ]
              }
              testID="input-field-email-test-id"
              value=""
            />
          </View>
          <View
            testID="password-input-signup-form-test-id"
          >
            <Text
              style={
                Object {
                  "color": "#3D3D3D",
                  "marginBottom": 3,
                }
              }
              testID="input-label-password-test-id"
            >
              Password
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "borderColor": "#9E9E9E",
                    "borderRadius": 8,
                    "borderWidth": 1,
                    "height": 40,
                    "padding": 10,
                  },
                  Object {
                    "borderColor": "#9E9E9E",
                  },
                ]
              }
              testID="input-field-password-test-id"
              value=""
            />
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FF4684",
                "borderRadius": 10,
                "elevation": 5,
                "padding": 15,
              }
            }
            testID="btn-submit-signup-test-id"
          >
            <Text
              style={
                Object {
                  "color": "#EDEDED",
                  "letterSpacing": 1,
                  "textTransform": "uppercase",
                }
              }
            >
              Sign Up
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
        testID="divider-signup-test-id"
      >
        <View
          style={
            Object {
              "alignSelf": "stretch",
              "backgroundColor": "#9E9E9E",
              "height": 1,
            }
          }
        />
        <Text
          style={
            Object {
              "backgroundColor": "#fff",
              "borderColor": "#9E9E9E",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#9E9E9E",
              "padding": 4,
              "position": "absolute",
              "textAlign": "center",
              "textTransform": "uppercase",
              "zIndex": 2,
            }
          }
        >
          OR
        </Text>
      </View>
      <View
        style={
          Object {
            "marginVertical": 60,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "alignItems": "center",
                "justifyContent": "space-around",
                "paddingHorizontal": 55,
              },
            ]
          }
          testID="auth-socials-signup-template-test-id"
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 100,
                  "borderWidth": 2,
                  "height": 50,
                  "justifyContent": "center",
                  "padding": 8,
                  "width": 50,
                },
                Object {
                  "borderColor": "#DB4437",
                },
              ]
            }
            testID="icon-btn-Google-test-id"
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "fontWeight": "bold",
                  },
                  Object {
                    "color": "#DB4437",
                  },
                ]
              }
            >
              G
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 100,
                  "borderWidth": 2,
                  "height": 50,
                  "justifyContent": "center",
                  "padding": 8,
                  "width": 50,
                },
                Object {
                  "borderColor": "#4267B2",
                },
              ]
            }
            testID="icon-btn-Facebook-test-id"
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "fontWeight": "bold",
                  },
                  Object {
                    "color": "#4267B2",
                  },
                ]
              }
            >
              f
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "borderRadius": 100,
                  "borderWidth": 2,
                  "height": 50,
                  "justifyContent": "center",
                  "padding": 8,
                  "width": 50,
                },
                Object {
                  "borderColor": "#0077B5",
                },
              ]
            }
            testID="icon-btn-LinkedIn-test-id"
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "fontWeight": "bold",
                  },
                  Object {
                    "color": "#0077B5",
                  },
                ]
              }
            >
              in
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "justifyContent": "center",
              },
            ]
          }
          testID="auth-footer-signup-template-test-id"
        >
          <Text
            style={
              Object {
                "color": "#9E9E9E",
                "marginHorizontal": 2,
              }
            }
          >
            Already a user?
          </Text>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="btn-footer-test-id"
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#9E9E9E",
                    "marginHorizontal": 2,
                  },
                  Object {
                    "textDecorationLine": "underline",
                    "textTransform": "uppercase",
                  },
                ]
              }
            >
              Login
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
